ui/prometheus-plugin/src/plugins/prometheus-time-series-query/query-editor-model.ts

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useState } from 'react';
+import { TextFieldProps } from '@mui/material';
+import { produce } from 'immer';
+import { OptionsEditorProps } from '@perses-dev/plugin-system';
+import { PrometheusTimeSeriesQuerySpec } from './time-series-query-model';
+
+export type PrometheusTimeSeriesQueryEditorProps = OptionsEditorProps<PrometheusTimeSeriesQuerySpec>;
+
+/**
+ * A hook for managing the `query` state in PrometheusTimeSeriesQuerySpec. Returns the `query` value, along with
+ * `onChange` and `onBlur` event handlers to the input. Keeps a local copy of the user's input and only syncs those
+ * changes with the overall spec value once the input is blurred to prevent re-running queries in the panel's preview
+ * every time the user types.
+ */
+export function useQueryState(props: PrometheusTimeSeriesQueryEditorProps) {
+  const { onChange, value } = props;
+
+  // Local copy of the query's value
+  const [query, setQuery] = useState(value.query);
+
+  // This is basically "getDerivedStateFromProps" to make sure if spec's value changes external to this component,
+  // we render with the latest value
```

Comment:
What kind of external change might cause this?

Reply:
Any component that's higher up in the tree also has access to getting/setting the spec state (going all the way up to the `PanelDrawer` itself). So the prop value that's getting passed in _could_ get changed higher up in the tree and become out of sync with our local state. In practice, hopefully that never happens (cause it's gonna make the state flows really weird).